js/packages/openinference-instrumentation-llama-index/src/instrumentation.ts

```diff
@@ -87,72 +98,160 @@ export class LlamaIndexInstrumentation extends InstrumentationBase<
-      (original: RetrieverQueryEngineQueryType): any => {
-        return function patchedQuery(
-          this: unknown,
-          ...args: Parameters<RetrieverQueryEngineQueryType>
-        ) {
-          const span = instrumentation.tracer.startSpan(`Query`, {
-            kind: SpanKind.INTERNAL,
-            attributes: {
-              [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
-                OpenInferenceSpanKind.CHAIN,
-              [SemanticConventions.INPUT_VALUE]: args[0].query,
-            },
-          });
-
-          const execContext = getExecContext(span);
-
-          const execPromise = safeExecuteInTheMiddle<
-            ReturnType<RetrieverQueryEngineQueryType>
-          >(
-            () => {
-              return context.with(execContext, () => {
-                return original.apply(this, args);
-              });
-            },
-            (error) => {
-              // Push the error to the span
-              if (error) {
-                span.recordException(error);
-                span.setStatus({
-                  code: SpanStatusCode.ERROR,
-                  message: error.message,
-                });
-                span.end();
-              }
-            },
-          );
-          const wrappedPromise = execPromise.then((result) => {
-            span.setAttributes({
-              [SemanticConventions.OUTPUT_VALUE]: result.response,
-            });
-            span.end();
-            return result;
-          });
-          return context.bind(execContext, wrappedPromise);
-        };
+      (original): any => {
+        return this.patchQueryMethod(
+          original,
+          moduleExports,
+          this
+        );
       },
     );
 
-    _isOpenInferencePatched = true;
+    this._wrap(
+      moduleExports.VectorIndexRetriever.prototype,
+      "retrieve",
+      (original): any => {
+        return this.patchRetrieveMethod(
+          original,
+          moduleExports,
+          this
+        );
+      },
+    )
 
+    _isOpenInferencePatched = true;
     return moduleExports;
   }
 
   private unpatch(moduleExports: typeof llamaindex, moduleVersion?: string) {
     this._diag.debug(`Un-patching ${MODULE_NAME}@${moduleVersion}`);
     this._unwrap(moduleExports.RetrieverQueryEngine.prototype, "query");
+    this._unwrap(moduleExports.VectorIndexRetriever.prototype, "retrieve");
 
     _isOpenInferencePatched = false;
   }
+
+  private patchQueryMethod(
+    original: typeof module.RetrieverQueryEngine.prototype.query,
+    module: typeof llamaindex,
+    instrumentation: LlamaIndexInstrumentation
+  ) {
+    return function patchedQuery(
+      this: unknown,
+      ...args: Parameters<typeof module.RetrieverQueryEngine.prototype.query>
+    ) {
+      const span = instrumentation.tracer.startSpan(`query`, {
+        kind: SpanKind.INTERNAL,
+        attributes: {
+          [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
+            OpenInferenceSpanKind.CHAIN,
```

Comment:
This is a question about llama index more so than the code here (also i know you didn't write this). Is this a chain because RetrieverQueryEngine.query kicks off a number of steps (retrieval, llm etc?)

Reply:
I believe so!